### PR TITLE
Stop Symbols from becoming Functions

### PR DESCRIPTION
**Pull request by cbm755**
There are various problems caused by Symbols automatically becoming Functions.  That is:
```py
>>> x = Symbol('x')
>>> f = Symbol('f')
>>> f(x)
f(x)
>>> srepr(_)
"Function('f')(Symbol('x'))"
```
It seems this exists so that `S('f(x)')` works.   See #3539 for discussion.  I got here from #9057 which is no longer failing per se.  But there are various other bug fixes here that were found by making this change.

This PR consists of some fixes cherry-picked from Diofant, with appreciaton.

#### Release notes

* core
  * Symbol `f` no longer automatically becomes a Function in `f(x)`.

### CODE REVIEW

**Comment by smichr** [comment]
Maybe you could give this a try:

```diff
diff --git a/sympy/core/function.py b/sympy/core/function.py
index 56bb4a4..7373363 100644
--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -441,8 +441,10 @@ def is_commutative(self):

     def _eval_evalf(self, prec):
         # Lookup mpmath function based on name
-        fname = self.func.__name__
         try:
+            if isinstance(self.func, UndefinedFunction):
+                raise AttributeError
+            fname = self.func.__name__
             if not hasattr(mpmath, fname):
                 from sympy.utilities.lambdify import MPMATH_TRANSLATIONS
                 fname = MPMATH_TRANSLATIONS[fname]
```

File: sympy/solvers/ode.py [hunk 1]
```diff
@@ -6934,11 +6934,12 @@ def _linear_2eq_order1_type7(x, y, t, r, eq):
...
         sol1 = dsolve(diff(x(t),t) - r['a']*x(t) - r['b']*sol2).rhs
     else:
-        x0, y0 = symbols('x0, y0')              #x0 and y0 being particular solutions
+        x0 = Function('x0')(t)    # x0 and y0 being particular solutions
+        y0 = Function('y0')(t)
         F = exp(Integral(r['a'],t))
         P = exp(Integral(r['d'],t))
         sol1 = C1*x0 + C2*x0*Integral(r['b']*F*P/x0**2, t)
-        sol2 = C1*y0 + C2(F*P/x0 + y0*Integral(r['b']*F*P/x0**2, t))
+        sol2 = C1*y0 + C2*(F*P/x0 + y0*Integral(r['b']*F*P/x0**2, t))
```
Comment (cbm755) [inline sympy/solvers/ode.py]: This is a nice example of the kind of "type errors" that creep in when Symbols can automatically become Functions.

**Comment by smichr** [comment]
`cls=Function` edits can be made to ode.py.

Also, how is E^y(x) parsed now? It was referred to in the unreviewed commit (f0e6ab7) you mentioned but I didn't see a test for it.

**Comment by cbm755** [comment]
Both on master and after this PR, it seems to do what I expect:
```py
>>> srepr(S('E^y(z)'))
"exp(Function('y')(Symbol('z')))"
```
I'll look into that a commit log a bit more.  At the very least, I will add the above as a test...

**Comment by cbm755** [comment]
BTW @smichr a long time ago you [suggested something above](https://github.com/sympy/sympy/pull/9064#issuecomment-76413809).  I've filed #15095 for that.

**Comment by asmeurer** [comment]
I would double check that all the changes to the parser are tested. I recall there not being a ton of tests for the parser.

**Comment by cbm755** [comment]
Ok, I went over f0e6ab7 and added a few tests.  It LGTM.

File: sympy/parsing/sympy_parser.py [hunk 2]
```diff
@@ -385,25 +389,28 @@ def _split_symbols(tokens, local_dict, global_dict):
...
-                    for char in symbol:
+                    tok_type = result[-2][1]  # Symbol or Function
+                    del result[-2:]  # Get rid of the call to Symbol
+                    for char in symbol[:-1]:
                         if char in local_dict or char in global_dict:
-                            # Get rid of the call to Symbol
-                            del result[-2:]
-                            result.extend([(NAME, "%s" % char),
-                                           (NAME, 'Symbol'), (OP, '(')])
+                            result.extend([(NAME, "%s" % char)])
```
Comment (smichr) [inline sympy/parsing/sympy_parser.py]: This is the only line that was added that is not hit by tests of `parsing` according to my analysis. The following hits it:
```
>>> t=transformations = standard_transformations + (split_symbols, implicit_multiplication,)
>>> parse_expr('yx', local_dict={'x':w},transformations=t)
w*y
```

**Comment by asmeurer** [comment]
Great to finally have this fixed.

**Pull request merged**